controllers/k8ssandra/k8ssandracluster_controller_test.go

```diff
@@ -2615,3 +2644,87 @@ func createMultiDcDseCluster(t *testing.T, ctx context.Context, f *framework.Fra
 	f.AssertObjectDoesNotExist(ctx, t, dc1Key, &cassdcapi.CassandraDatacenter{}, timeout, interval)
 	f.AssertObjectDoesNotExist(ctx, t, dc2Key, &cassdcapi.CassandraDatacenter{}, timeout, interval)
 }
+
+func testGenerationCheck(t *testing.T, ctx context.Context, f *framework.Framework, namespace string) {
+	require := require.New(t)
+
+	kc := &api.K8ssandraCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      "test",
+		},
+		Spec: api.K8ssandraClusterSpec{
+			Cassandra: &api.CassandraClusterTemplate{
+				ClusterName: "Not K8s_Compliant",
+				Datacenters: []api.CassandraDatacenterTemplate{
+					{
+						Meta: api.EmbeddedObjectMeta{
+							Name: "dc1",
+						},
+						K8sContext: f.DataPlaneContexts[1],
+						Size:       1,
+						DatacenterOptions: api.DatacenterOptions{
+							ServerVersion: "3.11.14",
+							StorageConfig: &cassdcapi.StorageConfig{
+								CassandraDataVolumeClaimSpec: &corev1.PersistentVolumeClaimSpec{
+									StorageClassName: &defaultStorageClass,
+								},
+							},
+							PodSecurityContext: &corev1.PodSecurityContext{
+								RunAsUser: pointer.Int64(999),
+							},
+							ManagementApiAuth: &cassdcapi.ManagementApiAuthConfig{
+								Insecure: &cassdcapi.ManagementApiAuthInsecureConfig{},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	err := f.Client.Create(ctx, kc)
+	require.NoError(err, "failed to create K8ssandraCluster")
+
+	verifyFinalizerAdded(ctx, t, f, kc)
+
+	verifySuperuserSecretCreated(ctx, t, f, kc)
+
+	verifyReplicatedSecretReconciled(ctx, t, f, kc)
+
+	verifySystemReplicationAnnotationSet(ctx, t, f, kc)
+
+	t.Log("check that the datacenter was created")
+	dcKey := framework.ClusterKey{NamespacedName: types.NamespacedName{Namespace: namespace, Name: "dc1"}, K8sContext: f.DataPlaneContexts[1]}
+	require.Eventually(f.DatacenterExists(ctx, dcKey), timeout, interval)
+
+	t.Log("update datacenter status to ready")
+	err = f.SetDatacenterStatusReady(ctx, dcKey)
+	require.NoError(err, "failed to set datacenter status ready")
+
+	verifyClusterReconcileFinished(ctx, t, f, kc)
+
+	// Modify the CassandraDatacenter hash to be some gibberish
```

Comment:
Yes, this is just a hack for the testing purpose. The idea is that the new version of k8ssandra-operator would create a different hash (since it wants to create different kind of CassandraDatacenter) which then wouldn't match the old one.

There's no other easy way to cause this in the test without causing Generation to change also (which we don't want in the test).